Add nest SDM API camera/doorbell events

allenporter opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Add binary_sensor for each event trait type supported by every device.  The sensor
is turned 'on' when pubsub messages are received, with an alarm to turn it back off after 30 seconds (arbitrary).  The pubsub messages for the device are filtered to just specific traits for each sensor, and ignores
messages that are too old and possibly no longer relevant.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/15512

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [X] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [X] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

allenporter committed:
Add nest SDM API binary_sensor for camera/doorbell

Add binary_sensor for each trait type supported by a device.  The sensor
is turned on when pubsub messages are received.  The pubsub messages for the
device are filtered to just specific traits for each sensor, and ignores
messages that are too old and possibly no longer relevant.

balloob commented on homeassistant/components/nest/binary_sensor_sdm.py:
```diff
@@ -0,0 +1,198 @@
+from google_nest_sdm.device import Device
+from google_nest_sdm.doorbell_traits import DoorbellChimeTrait
+from google_nest_sdm.event import (
+    CameraMotionEvent,
```
We should not show motion events as binary sensors. Instead, they should be fired as events of type `nest_event` and the pubsub -> HA events logic should be instantiated from `async_setup_entry` in `__init__.py`, it should not be tied to an entity. 

In a future PR device triggers can be offered to allow easy automating it. Example of a ZHA remote (also event based):

![image](https://user-images.githubusercontent.com/1444314/98167566-902ccb00-1ee9-11eb-8a97-39002d7a668a.png)

allenporter replied:
Thank you for the guidance and pointer for an example to follow.  I'm converting this to a draft while I update.

allenporter committed:
Update event tests

allenporter committed:
Fire events on doorbell or camera events

allenporter committed:
Revert binary_sensor in nest api

allenporter committed:
Rename binary sensor without legacy bridge

balloob commented on homeassistant/components/nest/__init__.py:
```diff
@@ -173,17 +188,16 @@ def handle_event(self, event_message: EventMessage):
-        async_dispatcher_send(self._hass, SIGNAL_NEST_UPDATE)
+            for (event, data) in events.items():
+                if event in EVENT_TRAITS:
+                    self._hass.bus.fire(NEST_EVENT, data)
```
Is this method being invoked in async or sync context? You are using `async_dispatcher_send`, so I am assuming async, in that case:
```suggestion
                    self._hass.bus.async_fire(NEST_EVENT, data)
```

If it's sync, it should be `dispatcher_send`. 

Can you confirm that `resource_update_events` includes the `device_id` of the device that fired the event? It's something I added to the documentation last week <https://developers.home-assistant.io/docs/integration_events>. It will help with writing automations as we can listen for `nest_event` with `device_id: <thermostat ID>` to get all events of the thermostat.

allenporter replied:
Thanks for the pointer. 

I've updated this to create a separate data message, independent from the SDM library.  I have two questions here:
1) Since this is from a pubsub background thread, i am pretty sure this is not async.  However, it only passes the tests in async, as the callback isnt fired by the time the test has run.  I was unable to find a lot of tests that exercise the blocking dispatcher_send() or bus.fire() that also had tests.  Have a suggestion for how to block until all events have run?
2) Should this move device updates to use the bus instead of the dispatcher? It'd basically just make both ways to listen to events uniform, and it could also only update specific devices that changed only.

Thanks

balloob commented on tests/components/nest/test_events.py:
```diff
@@ -43,16 +45,15 @@ async def test_event_update_event(hass):
-
     count = 0
 
     def handle_event(event):
```
This test runs a little faster if you annotate this method with `@callback`, as it won't run this method in a thread. We also have a test helper: `events = async_capture_events(hass, "nest_event")` that you can import from `tests.common`. `events` will be a list with all events fired. You would then do `assert len(events) == 0` etc.

balloob reviewed (approved):
Excited about events! 🎉

axwell commented:
looking forward to test this, in what version will be available ?

allenporter committed:
Add device and type to event data.

allenporter committed:
Remove unused event

allenporter committed:
Remove unused variable

allenporter commented:
This is ready for another review.  I'm still having trouble with the sync vs async in tests.

allenporter committed:
Update subscriber test methods to new API

Catch up with latest merge and update renamed methods.

allenporter committed:
Fix event message to use HA device ids

Fire messages on the event bus using Home Assistant device ids, instead
of nest API specific device ids.  This adds an extra step to
resolve a device id when an event message is received.

allenporter commented:
This is ready for re-review.  I've updated this to reflect fully a async subscriber andI fixed an issue where the device ids were in the nest API format, and not home assistant device ids, in order to make this work better with triggers.

balloob commented on homeassistant/components/nest/__init__.py:
```diff
@@ -169,21 +184,41 @@ def __init__(self, hass: HomeAssistant):
+            if not device_entry:
+                _LOGGER.debug("Ignoring event for unregistered device '%s'", device_id)
+                return
+            for event in events.keys():
```
```suggestion
            for event in events:
```

balloob commented on homeassistant/components/nest/__init__.py:
```diff
@@ -169,21 +184,41 @@ def __init__(self, hass: HomeAssistant):
+            # Send a signal to refresh state of all listening devices.
+            async_dispatcher_send(self._hass, SIGNAL_NEST_UPDATE)
         events = event_message.resource_update_events
         if events:
```
Consider turning this around: `if not events: return`

balloob commented on tests/components/nest/test_events.py:
```diff
@@ -0,0 +1,237 @@
+    await hass.async_block_till_done()
+
+    assert len(events) == 1
+    assert {
+        "device_id": entry.device_id,
+        "type": "CameraSound",
+    } == events[0].data
```
This is really minor but the expected value should be on the right of `==`
```suggestion
    assert events[0].data == {
        "device_id": entry.device_id,
        "type": "CameraSound",
    }
```

allenporter committed:
Update homeassistant/components/nest/__init__.py

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

allenporter committed:
Update nest code style based on PR feedback

allenporter committed:
Merge branch 'binary_sensor' of github.com:allenporter/home-assistant-core into binary_sensor
